### PR TITLE
Add experimental support for Thread Sanitizer

### PR DESCRIPTION
**Pull request by AnnaZaks**
Thread the -sanitize=thread option through the front end and fix a bunch of races it reported.

### CODE REVIEW

**Comment by AnnaZaks** [comment]
@rjmccall, @gribozavr please, take a look!
@swift-ci Please test

**Comment by gribozavr** [comment]
@AnnaZaks Library changes LGTM!

File: include/swift/Runtime/Concurrent.h [hunk 1]
```diff
@@ -297,7 +297,7 @@ template <class EntryTy> class ConcurrentMap {
 
       // Try to set the edge to the new node.
       if (std::atomic_compare_exchange_strong_explicit(edge, &node, newNode,
-                                                  std::memory_order_release,
+                                                  std::memory_order_acq_rel,
```
Comment (rjmccall) [inline include/swift/Runtime/Concurrent.h]: Okay, but I'm going to link to [C++ DR 2445](http://cplusplus.github.io/LWG/lwg-active.html#2445) in protest. :)

File: lib/IRGen/GenMeta.cpp [hunk 2]
```diff
@@ -1075,6 +1075,9 @@ void irgen::emitLazyCacheAccessFunction(IRGenModule &IGM,
   // FIXME: Technically should be "consume", but that introduces barriers in the
   // current LLVM ARM backend.
   auto load = IGF.Builder.CreateLoad(cache);
+  if (IGM.Opts.Sanitize == SanitizerKind::Thread)
+    load->setOrdering(llvm::AtomicOrdering::Acquire);
+
```
Comment (rjmccall) [inline lib/IRGen/GenMeta.cpp]: Please leave a comment like:

  // Make this barrier explicit when building for TSan to avoid false positives.

**Comment by rjmccall** [comment]
Otherwise, runtime/IRGen changes LGTM.

**Comment by gparker42** [comment]
Can we separate the "enable tsan" changes from the "fix stuff detected by tsan" changes?

File: lib/Option/SanitizerOptions.cpp [hunk 3]
```diff
@@ -36,23 +40,50 @@ SanitizerKind swift::parseSanitizerArgValues(const llvm::opt::Arg *A,
...
+    if (kind == SanitizerKind::None) {
       Diags.diagnose(SourceLoc(), diag::error_unsupported_option_argument,
-        A->getOption().getName(), A->getValue(i));
+        A->getOption().getPrefixedName(), A->getValue(i));
       return kind;
     }
+
+    // Currently, more than one sanitizer cannot be enabled at the same time.
+    if (pKind != SanitizerKind::None && pKind != kind) {
+      SmallVector<char, 128> pb;
```
Comment (jrose-apple) [inline lib/Option/SanitizerOptions.cpp]: Nitpick: SmallString?

File: lib/Option/SanitizerOptions.cpp [hunk 4]
```diff
@@ -36,23 +40,50 @@ SanitizerKind swift::parseSanitizerArgValues(const llvm::opt::Arg *A,
...
   // Find the sanitizer kind.
-  // TODO: Add support for dealing with multiple sanitizers.
+  SanitizerKind pKind = SanitizerKind::None;
   for (int i = 0, n = A->getNumValues(); i != n; ++i) {
-    const char *Value = A->getValue(i);
-    if (StringRef(Value).equals("address")) {
-      kind = SanitizerKind::Address;
-    } else {
+    kind =
+    llvm::StringSwitch<SanitizerKind>(A->getValue(i))
```
Comment (jrose-apple) [inline lib/Option/SanitizerOptions.cpp]: Weird indentation here.

**Comment by AnnaZaks** [comment]
@gparker42 They are all separate commits, which you can view individually.

**Comment by AnnaZaks** [comment]
@swift-ci Please test

**Comment by AnnaZaks** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
Awesome!

**Pull request merged**